crates/task-impls/src/transactions.rs

```diff
@@ -235,68 +130,79 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
     }
 
     #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Transaction Handling Task", level = "error")]
-    async fn wait_for_transactions(&self) -> Option<Vec<TYPES::Transaction>> {
+    async fn wait_for_block(
+        &self,
+    ) -> Option<(AvailableBlockData<TYPES>, AvailableBlockHeaderInput<TYPES>)> {
         let task_start_time = Instant::now();
 
-        // TODO (Keyao) Investigate the use of transaction hash
-        // <https://github.com/EspressoSystems/HotShot/issues/1811>
-        // let parent_leaf = self.parent_leaf().await?;
-        // let previous_used_txns = match parent_leaf.tarnsaction_commitments {
-        //     Some(txns) => txns,
-        //     None => HashSet::new(),
-        // };
+        let last_leaf = self.consensus.read().await.get_decided_leaf();
+        let mut latest_block: Option<(
+            AvailableBlockData<TYPES>,
+            AvailableBlockHeaderInput<TYPES>,
+        )> = None;
+        while task_start_time.elapsed() < self.api.propose_max_round_time()
+            && latest_block.as_ref().map_or(true, |(data, _)| {
+                data.block_payload.get_transactions(&data.metadata).len()
+                    < self.api.min_transactions()
+            })
+        {
+            let mut available_blocks = match self
+                .builder_client
+                .get_available_blocks(last_leaf.block_header.payload_commitment())
+                .await
+            {
+                Ok(blocks) => blocks,
+                Err(err) => {
+                    error!(%err, "Couldn't get available blocks");
+                    continue;
+                }
+            };
+
+            available_blocks.sort_by_key(|block_info| block_info.block_size);
```

Comment:
Should we be sorting by the fee, intead of size?

Reply:
I was thinking of trying to include the most transactions instead, but it doesn't necessarily follow from bigger block size. Sorting by fee makes more sense, I'll switch to that.